Issue62 Fix: USB Cable Connection Exceptions

kb1lqc opened the pull request:
This PR fixes the complaints in #62 which essentially deal with the USB connection. Both scenarios are now caught and handled.

- Removing USB during `Proxy` operation
```
C:\Python27\python.exe C:/Users/Bryce/Documents/GitHub/faradayrf-kb1lqc/Proxy/proxy.py
2017-04-30 11:10:29,854 - Proxy - INFO - Starting proxy server
2017-04-30 11:10:29,903 - Proxy - INFO - Starting Thread For Unit: KB1LQC-2
2017-04-30 11:10:29,903 - Proxy - INFO - Starting uart_worker thread
2017-04-30 11:10:29,904 - Proxy - INFO - Starting Thread For Unit: KB1LQC-1
2017-04-30 11:10:29,904 - Proxy - INFO - Starting uart_worker thread
2017-04-30 11:10:31,104 - root - ERROR - USB cable likely disconnected!
2017-04-30 11:10:31,104 - root - ERROR - call to ClearCommError failed
```

- Starting `Proxy` with any USB connection missing
```
C:\Python27\python.exe C:/Users/Bryce/Documents/GitHub/faradayrf-kb1lqc/Proxy/proxy.py
2017-04-30 11:04:46,812 - Proxy - INFO - Starting proxy server
2017-04-30 11:04:46,838 - root - ERROR - Check USB cable connection!
2017-04-30 11:04:46,838 - root - ERROR - could not open port 'COM42': WindowsError(2, 'The system cannot find the file specified.')
```

## 7-6-2017 UARTstack Comment

@reillyeon it appears @kb1lqd doesn't want to remove the [UART service layer 2 object](https://github.com/FaradayRF/Faraday-Software/pull/195#discussion_r125560959). He does make a good point that the UART code is SO old that we really should consider completely rewriting it one day. There are a ton of brute force timer sleeps and such in it that I have no doubt will become a bottleneck one day and force us to reconsider the architecture of the class itself. This will probably happen in the near future with data applications on the horizon. **Therefore we're punting this concern to a later issue ticket**

kb1lqc committed:
Added Try/Except Statement for USB Disconnect

When the USB cable is disconnected this try/except statement now exits
python and displays a helpful message suggesting that the USB cable has
been disconnected. This exits all of Proxy as a consequesce so it's pretty
drastic but seems reasonable. It's not guaranteed to be the USB disconnect
everytime but most likely will be. Addresses a portion of Issue #62

kb1lqc committed:
Added Try/Except for Unconnected USB Cable

Addresses Issue #62 complaint that no USB connection upon start of Proxy
results in an error. this now catches the exception and provides a helpful
error message to the user.

kb1lqc committed:
Fixed PEP8 Failures

I was bad an forgot to check pytest prior to pushing to the repo. Fixed
too many lines errors.

kb1lqc commented:
I know I used a different method of sys.exit() in this fix. Instead of importing all of sys I just imported exit() from sys. Not sure if I should keep it consistent. This seemed like less of a footprint.

reillyeon commented:
I prefer not importing individual functions from a module because when you see `sys.exit` you know what it means whereas `exit` could mean anything. Also, `sys.exit(1)` since these are failures. To clarify my earlier comment: I don't believe it is worthwhile to define failure codes other than 0 for success, 1 for error.

kb1lqc committed:
Updated Exit Code to 1

Using exit code of 1 to indicate error state.

kb1lqc committed:
Using sys.exit

Importing the entire module for clarity with sys.exit() use.

kb1lqc commented:
@reillyeon fixed and yes that actually makes more sense. Simply using exit code of 1 for anything incorrectly exiting.

reillyeon commented on faraday/uart/layer_2_protocol.py:
```diff
@@ -15,14 +15,20 @@
 logger = logging.getLogger('UARTStack')
 
 
 class layer_2_object(object):
```
Can we just remove this class entirely? It looks like it doesn't serve any purpose than to hold a layer_2_protocol object.

kb1lqc replied:
I beleive @kb1lqd implemented it as an abstraction. I can dig into it too.

kb1lqc replied:
@reillyeon it appears @kb1lqd used this in [layer 2 service](https://github.com/kb1lqc/Faraday-Software/blob/fec6feae17b3e843459d6582d7040354bba2c51a/faraday/uart/layer_2_service.py#L26) I think it probably can be removed if this line is updated. Am I correct Brent?

kb1lqd replied:
I seem to only use it locally in layer_2_object() class. I agree that this may be a bit convoluted and unneeded but the protocol object itself certainly is used. I believe it's best to leave it here for now and have a seperate IT/PR to update the lower level UART code for cleanliness/optimization. We're going to need to likely do this anyways with the improvement of generic data operation on the near term horizon.

reillyeon commented on faraday/uart/layer_2_protocol.py:
```diff
@@ -67,9 +73,14 @@ def run(self):
+                except serial.SerialException as e:
+                    logging.error("USB cable likely disconnected!")
+                    logging.error(e)
+                    sys.exit(1)  # Sys.exit(1) is an error
```
Why not call `self.abort()` here instead? I'm concerned that adding `sys.exit` calls too deeply layered in the code will make it difficult for us to decide later what errors are fatal or not.

el-iso replied:
@reillyeon is abort() from "os"? Are you suggesting that it would be better because it triggers a core dump? Just trying to educate myself on this subject....

reillyeon replied:
`abort()` is a method on this class that sets `self.enabled` to `false` which will cause this loop to terminate and the thread to exit.

el-iso replied:
Oh, that was dumb of me. I didn't even think to look for the implementation here... Thanks

kb1lqc committed:
Merge branch 'master' into issue62-2

PUlling in the latest master which includes actually launching from a pip
installed proxy instead of the old manual way. This is a massive update.

kb1lqc committed:
Removed sys.exit() use and used self.abort()

Listened to @reillyeon and use self.abort() instead of sys.exit(). Also,
cleaned up error logging for these functions.

kb1lqc committed:
Updated UARTstack logging to WARNING

kb1lqc committed:
Cleaned Up Logging

Used logger instead of logging and fixed up the flow of errors.

kb1lqc committed:
Removed unused variable

pytest complained about an unused error 'e' which provides little info so
I removed it.

kb1lqc committed:
Added logger information to layer_2_service.py

Added basic logger error functions to this class.

kb1lqc committed:
Updated logging in UART services

Drastically improved UART service monitoring and error/info display in
UART layers.

reillyeon commented on faraday/uart/layer_2_protocol.py:
```diff
@@ -15,19 +15,28 @@
-        self.serial_physical_obj = layer_2_protocol(port, baud, timeout)
+        # Start up serial object and if it fails exit the program
+        try:
+            self.serial_physical_obj = layer_2_protocol(port, baud, timeout)
```
Move this try/catch block into the layer_2_protocol constructor so it is clear what SerialException is being caught (the one that might be thrown by the serial.Serial constructor).

kb1lqc replied:
Just pushed an update addressing this comment. Is this what you wanted?

reillyeon replied:
Yes.

reillyeon reviewed (approved):
This looks good to me with one nitpick about where to place a try/catch block.

kb1lqc committed:
Removed pytest warnings

Fixed up all pytest warnings including commenting out some unused code.

kb1lqc commented:
Updated this branch with the latest master code with a merge. All seems well. Used `pip install -e .` to install and edit the later updates to logging and error handling.

In addition to ensuring correct exceptions when USB cable is removed I also updated the logging use in the UART services. Below is an example of a UART communications error that results in an unsuccessful struct.unpack(). I now use appropriate warnings level and log the data in text format.

![image](https://user-images.githubusercontent.com/331540/27849992-44489c5e-6106-11e7-9c8e-0fa676362cd8.png)

kb1lqc committed:
Moved try/except statement to layer 2 protocol

Per @reillyeon I moved the try/except statement into the layer 2 protocol
calss constructor.

kb1lqc committed:
Fixed pytest line warnings

Messed up line spacing when I moved the try/except. This commit fixes that
warning from pytest.

kb1lqc commented:
This PR is ready to merge pending @kb1lqd thoughts on removing the layer 2 object

kb1lqd commented:
Testing indicates some questions.

I used this PR to write some quick batch scripts: [faraday-bat.zip](https://github.com/FaradayRF/Faraday-Software/files/1125969/faraday-bat.zip)

# Unit disconnections:

2 units were connected and I connected telemetry to the unit-1 (COM8). Disconnecting unit-2 (COM112) showed the intended error message and telemetry continued to be saved from the remaining connected unit.

![image](https://user-images.githubusercontent.com/339144/27884057-ba657408-6187-11e7-9743-53faf0809cbe.png)

# Starting proxy with one unit disconnected
I kept proxy with option `--number 2` but disconnected the COM112 unit and started proxy. Proxy refuses to start:

![image](https://user-images.githubusercontent.com/339144/27884147-19f0e5e2-6188-11e7-955c-fb7b453c65e2.png)

@kb1lqc I thought that the intended operation was that proxy WILL start if one of the units is missing? What is the intended operation?

kb1lqd commented on etc/faraday/loggingConfig.ini:
```diff
@@ -36,7 +36,7 @@ qualname=SimpleConfig
 
 [logger_UARTStack]
-level=INFO
+level=WARNING
```
OK

kb1lqd reviewed (changes requested):
Looks good but my only request is to clarify operation when proxy is started expecting multiple units but only a single unit is connected? I seem to experience unintended failure of proxy booting in this condition.

kb1lqc commented:
Ok @kb1lqd I'll look at it in the morning. Multiple units is something I didn't eve consider tbh. Good catc

kb1lqc committed:
Merge remote-tracking branch 'upstream/master' into issue62-2

kb1lqc committed:
Fails gracefully now with one unit connected multi

When configured for multiple radios but only one is connected then thecode
simply excepts and continues on with a try/except statement in
proxy:main(). Also cleaned up the node name for this section of code with
a node variable.

kb1lqc committed:
Fixed whitespace pytest PEP8 warning

kb1lqc commented:
@kb1lqd I commented in #62 with the additions that make multiple unit configurations work better and fail gracefully. I believe this is all you need to be OK with approving and merging? Please do so if you are OK with these changes!

kb1lqd commented:
@kb1lqc Thanks for the update, just to clarify: I still cannot start proxy with only one unit connected but proxy expecting 2 units.
![image](https://user-images.githubusercontent.com/339144/27937979-cceb5096-626f-11e7-9de1-6b54d2e88861.png)

I don't see the text failure that you implemented here: https://github.com/FaradayRF/Faraday-Software/pull/195/commits/37b0100b928f5be40554e554b7d8c6794b4e3db9#diff-a9d2cba55827323291e9d0c49fd6f720R818

All in all it seems to be working but this one case isn't handled. I'd expect proxy to connect to all available units connected and ignore those not connected but configured to connect to by error'ing on them but not blocking operation. Is this not the expected behavior?

I'm cool with approving this if we acknowledge that you EXPECT proxy to not operate when:
- Config specifies 2 units connected
- Only a single unit connected when proxy boots

@kb1lqc Is https://github.com/FaradayRF/Faraday-Software/pull/235 suppose to fix my question above?

kb1lqd commented on faraday/proxy.py:
```diff
@@ -811,7 +811,11 @@ def main():
+                node = str(values["callsign"] + "-" + values["nodeid"])
+                unitDict[node] = layer_4_service.faraday_uart_object(str(values["com"]), int(values["baudrate"]), int(values["timeout"]))
+            except:
+                logger.error('Could not connect to {0} on {1}'.format(node, values["com"]))
```
OK good. Seems to work.

kb1lqd reviewed (approved):
Latest commit seems to address my concerns.
